### DISCUSSION TITLE
Convert and Orient Meshes from Morphologist 2021 and BV 5.0.4 to 3D Slicer 5.2.2

### DISCUSSION THREAD

**Discussion by Papagif66** [brainvisa/brainvisa.github.io#181] [Q&A] [answered]
Hi, et bonjour Denis (sure you'll be the one to answer :-) )
I've run Morphologist 2021 from Branvisa 5.0.4 and need to import the results (meshes and graph) into Slicer 5.2.2.
Starting with Lhemi and using Aims Mesh Converter and PLY mesh, I've got a mesh that is read by Slicer.
But I have an orientation problem that I can't solve.
I guess this is because the transform that is in the gii.minf file of the .gii input file is still in the ply.minf file of the .ply file, and that Slicer can't read the minf file, but only the .ply file.
Any ideas how to solve this?
Best,
Eric
![6015e601-03c6-4111-a709-0a90301e7683](https://github.com/user-attachments/assets/8d91cae6-7ae7-4d29-9fdc-33abfb22040a)

**Comment by denisri** [comment, marked as answer]
Hi Eric,
This is because AIMS/BrainVisa/Anatomist employ different coordinates systems conventions from 3D Slicer.
You can read https://brainvisa.info/aimsdata/user_doc/coordinates_systems.html
I don't know which coord systems 3D Slicer is using (input image disk voxels orientation ? Scanner coords ? A reference space like MNI.ICBM ?)
If you can find out, you can resample the data (mesh, volume, or even sulci graph) using the command AimsApplyTransform, but determining the correct transformation will be the most difficult part.
I realize that we have not provided an easy tool to get a transformation to voxels disk space in millimeters. We should add that one day when we can find a minute for that..
Denis

**Comment by denisri** [comment]
Here is a little script which extracts the transformation to disk orientation from an image.
You must use it within the brainvisa container in order to have the libraries available. It reads an image header, and writes a transformation file (.trm) to the disk oriented space. Then you can use AimsApplyTransform to transform/resample any object (image, mesh, sulci graph etc) to this space.
```
#!/usr/bin/env python

from soma import aims
import numpy as np


def aims_to_disk(header):
    '''
    Extracts the transformation from the AIMS referential of the object header
    to the disk storage space, in mm.

    This transformation is the inverse of the "storage_to_memory" property,
    plus conversion in mm space (storage_to_memory is in voxels).
    '''

    s2mo = header.get('storage_to_memory')
    if s2mo is None:
        return None

    s2m_vox = aims.AffineTransformation3d(s2mo)
    m2s_vox = s2m_vox.inverse()
    vs = [1., 1., 1.]
    vso = header.get('voxel_size')
    if vso is not None:
        vs = vso[:3]

    # get s2m in mm
    tvs1 = aims.AffineTransformation3d()
    tvs1.np[np.diag_indices(3)] = 1. / np.asarray(vs)

    vss = m2s_vox.transformVector(vs)  # vs in disk orientation
    tvs2 = aims.AffineTransformation3d()
    tvs2.np[np.diag_indices(3)] = np.abs(vss)

    m2s = tvs2 * m2s_vox * tvs1

    return m2s


if __name__ == '__main__':

    import argparse
    parser = argparse.ArgumentParser(
        description='Extracts the transformation from the AIMS referential of '
        'the object header to the disk storage space, in mm.')
    parser.add_argument(
        '-i', '--input',
        help='object to extract the transformation from. Normally a 3D MRI '
        'volume, but it may also be a mesh or a graph, if the voxels '
        'orientation information of the source MRI has been copied to the '
        'object header.')
    parser.add_argument(
        '-o', '--output', help='output transformation file (.trm)')
    args = parser.parse_args()

    in_file = args.input
    out_file = args.output

    f = aims.Finder()
    if not f.check(in_file):
        raise ValueError('the input file cannot be parsed')

    header = f.header()
    m2s = aims_to_disk(header)
    if m2s is None:
        raise ValueError('the transformation to disk space could not be found '
                         'and extrracted from the object header')
    aims.write(m2s, out_file)
```
Denis

**Comment by Papagif66** [comment]
H Denis (I knew it :-) ),
Thanx for your mail an script.
I'll try this and let you know.
Best,
Eric
